src/changelog.ts

```diff
@@ -192,42 +187,137 @@ function stringifyLinkReferenceDefinitions(
+    releases.length > 0 ? '\n' : ''
+  }`;
+}
+
+/**
+ * Get a string of unreleased link reference definition.
+ *
+ * @param repoUrl - The URL for the GitHub repository.
+ * @param tagPrefix - The prefix used in tags before the version number.
+ * @param releases - The releases to generate link definitions for.
+ * @param [versionBeforePkgRename] - A version string of the package before being renamed.
+ * @param [tagPrefixBeforePkgRename] - A tag prefix string of the package before being renamed.
+ * @returns A unreleased link reference definition string.
+ */
+function getUnreleasedLinkReferenceDefinition(
+  repoUrl: string,
+  tagPrefix: string,
+  releases: ReleaseMetadata[],
+  versionBeforePkgRename?: string,
+  tagPrefixBeforePkgRename?: string,
+): string {
+  // A list of release versions in descending SemVer order
+  const descendingSemverVersions = releases
+    .map(({ version }) => version)
+    .sort((a: Version, b: Version) => {
+      return semver.gt(a, b) ? -1 : 1;
+    });
+  const latestSemverVersion = descendingSemverVersions[0];
+  const hasReleases = descendingSemverVersions.length > 0;
+  // if there is a package renamed, the tag prefix before the rename will be considered for compare
+  // [Unreleased]: https://github.com/ExampleUsernameOrOrganization/ExampleRepository/compare/test@0.0.2...HEAD
+  const tagPrefixToCompare =
+    tagPrefixBeforePkgRename && versionBeforePkgRename === latestSemverVersion
+      ? tagPrefixBeforePkgRename
+      : tagPrefix;
+
+  return `[${unreleased}]: ${
+    hasReleases
+      ? getCompareUrl(
+          repoUrl,
+          `${tagPrefixToCompare}${latestSemverVersion}`,
+          'HEAD',
+        )
+      : withTrailingSlash(repoUrl)
+  }`;
+}
+
+/**
+ * Get a list of release link reference definitions.
+ *
+ * @param repoUrl - The URL for the GitHub repository.
+ * @param tagPrefix - The prefix used in tags before the version number.
+ * @param releases - The releases to generate link definitions for.
+ * @param [versionBeforePkgRename] - A version string of the package before being renamed.
+ * @param [tagPrefixBeforePkgRename] - A tag prefix string of the package before being renamed.
+ * @returns A list of release link reference definitions.
+ */
+function getReleaseLinkReferenceDefinitions(
+  repoUrl: string,
+  tagPrefix: string,
+  releases: ReleaseMetadata[],
+  versionBeforePkgRename?: string,
+  tagPrefixBeforePkgRename?: string,
+): string[] {
+  const chronologicalVersions = releases.map(({ version }) => version);
+  let tagPrefixToCompare = tagPrefix;
+  const releaseLinkReferenceDefinitions = releases.map(({ version }) => {
+    let diffUrl;
+    // once the version matches with original version, rest of the lines in changelog will be assumed as migrated tags
+    if (tagPrefixBeforePkgRename && versionBeforePkgRename === version) {
+      tagPrefixToCompare = tagPrefixBeforePkgRename;
+    }
```

Comment:
Sure. As the variables tagPrefixBeforePkgRename versionBeforePkgRename have been renamed. I'll commit this change along with rename.

Reply:
Alright. I guess the benefit of defining the variable inside the loop is that it creates a smaller chunk of code that can be understood on its own and refactored more easily in the future if necessary. This is minor, however.

Reply:
Ah. I see what you're saying.

What do you think about using a `for` loop rather than `map` then? I feel like this fact — that we _want_ a variable whose state persists outside of the loop rather than getting reset — is important and shouldn't get lost. By including it in the `map` I wonder if it would get glossed over, since the primary goal of the `map` is to build the string. But if we use a `for` loop we're intentionally using a more procedural pattern and thus we're communicating that there are multiple things we want the reader to pay attention to.